controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java

```diff
@@ -0,0 +1,242 @@
+                OptionalInt majorVersion, ApplicationMetrics metrics, Optional<String> pemDeployKey, OptionalLong projectId,
+                boolean internal, Collection<Instance> instances) {
+        this.id = Objects.requireNonNull(id, "id cannot be null");
+        this.createdAt = Objects.requireNonNull(createdAt, "instant of creation cannot be null");
+        this.deploymentSpec = Objects.requireNonNull(deploymentSpec, "deploymentSpec cannot be null");
+        this.validationOverrides = Objects.requireNonNull(validationOverrides, "validationOverrides cannot be null");
+        this.change = Objects.requireNonNull(change, "change cannot be null");
+        this.outstandingChange = Objects.requireNonNull(outstandingChange, "outstandingChange cannot be null");
+        this.deploymentIssueId = Objects.requireNonNull(deploymentIssueId, "deploymentIssueId cannot be null");
+        this.ownershipIssueId = Objects.requireNonNull(ownershipIssueId, "ownershipIssueId cannot be null");
+        this.owner = Objects.requireNonNull(owner, "owner cannot be null");
+        this.majorVersion = Objects.requireNonNull(majorVersion, "majorVersion cannot be null");
+        this.metrics = Objects.requireNonNull(metrics, "metrics cannot be null");
+        this.pemDeployKey = Objects.requireNonNull(pemDeployKey, "pemDeployKey cannot be null");
+        this.projectId = Objects.requireNonNull(projectId, "projectId cannot be null");
+        this.internal = internal;
+        this.instances = instances.stream().collect(Collectors.toUnmodifiableMap(instance -> instance.id().instance(),
+                                                                                 instance -> instance));
+    }
+
+    /** Returns an aggregate application, from the given instances, if at least one. */
+    public static Optional<Application> aggregate(List<Instance> instances) {
+        if (instances.isEmpty())
+            return Optional.empty();
+
+        Instance base = instances.stream()
+                                 .filter(instance -> instance.id().instance().isDefault())
+                                 .findFirst()
+                                 .orElse(instances.iterator().next());
+
+        return Optional.of(new Application(TenantAndApplicationId.from(base.id()), base.createdAt(), base.deploymentSpec(),
+                                           base.validationOverrides(), base.change(), base.outstandingChange(),
+                                           base.deploymentJobs().issueId(), base.ownershipIssueId(), base.owner(),
+                                           base.majorVersion(), base.metrics(), base.pemDeployKey(),
+                                           base.deploymentJobs().projectId(), base.deploymentJobs().deployedInternally(), instances));
+    }
+
+    /** Returns an old Instance representation of this and the given instance, for serialisation. */
+    public Instance legacy(InstanceName instance) {
```

Comment:
Temporary `Instance` creator. Uses application level data from this and instance level data from the instance to represent.